packages/core/src/network/heartbeat.ts

```diff
@@ -38,69 +63,113 @@ export default class Heartbeat {
+    log('ping', destination.toB58String())
 
     const challenge = randomBytes(16)
     const expectedResponse = Heartbeat.calculatePingResponse(challenge)
 
+    let pingResponse: Uint8Array[] | undefined
+
     try {
-      const pingResponse = await this.sendMessage(id, this.protocolHeartbeat, challenge, true, {
-        timeout: HEARTBEAT_TIMEOUT
+      pingResponse = await this.sendMessage(destination, this.protocolHeartbeat, challenge, true, {
+        timeout: this.config.heartbeatDialTimeout,
+        signal
       })
+    } catch (err) {
+      log(`Connection to ${destination.toB58String()} failed: ${err?.message}`)
+      return {
+        destination,
+        lastSeen: -1
+      }
+    }
+
+    if (pingResponse == null || pingResponse.length != 1 || !u8aEquals(expectedResponse, pingResponse[0])) {
+      log(`Mismatched challenge. Got ${u8aToHex(pingResponse[0])} but expected ${u8aToHex(expectedResponse)}`)
+      await this.hangUp(destination)
 
-      if (pingResponse == null || pingResponse.length == 0 || !u8aEquals(expectedResponse, pingResponse[0])) {
-        log(`Mismatched challenge. ${pingResponse}`)
-        await this.hangUp(id)
-        return false
+      return {
+        destination,
+        lastSeen: -1
       }
+    }
 
-      log('ping success to', id.toB58String())
-      return true
-    } catch (e) {
-      log(`Connection to ${id.toB58String()} failed: ${JSON.stringify(e)}`)
-      return false
+    return {
+      destination,
+      lastSeen: Date.now()
     }
   }
 
-  private async checkNodes(): Promise<void> {
-    const thresholdTime = Date.now() - HEARTBEAT_INTERVAL
+  /**
+   * Performs a ping request to all nodes who were not seen since the threshold
+   */
+  protected async checkNodes(): Promise<void> {
+    const thresholdTime = Date.now() - this.config.heartbeatThreshold
     log(`Checking nodes since ${thresholdTime} (${new Date(thresholdTime).toLocaleString()})`)
 
-    const toPing = this.networkPeers.pingSince(thresholdTime)
+    const abort = new AbortController()
+
+    let finished = false
+
+    setTimeout(() => {
+      if (!finished) {
+        abort.abort()
+      }
+    }, this.config.heartbeatRunTimeout)
+
+    // Create an object that describes which work has to be done
+    // by the workers, i.e. the pingNode code
+    const pingWork = this.networkPeers
+      .pingSince(thresholdTime)
+      .map<[destination: PeerId, signal: AbortSignal]>((peerToPing: PeerId) => [peerToPing, abort.signal])
+
+    const pingResults = await nAtATime(this._pingNode, pingWork, this.config.maxParallelHeartbeats)
 
-    const doPing = async (): Promise<void> => {
-      await this.networkPeers.ping(toPing.pop(), async (id: PeerId) => await this.pingNode(id))
+    finished = true
+
+    for (const pingResult of pingResults) {
+      // Filter unexpected network errors
+      if (pingResult instanceof Error) {
+        continue
+      }
+      this.networkPeers.updateRecord(pingResult)
```

Comment:
If we keep aborting the previous jobs using an additional timeout, we will need a mutex here to protect the `networkPeers`.

Reply:
The idea is to only abort `nAtATime`, so this code will be executed only after `nAtATime` returned.